### ISSUE TITLE
Ability to compare Units (or possibly other entities)

### ISSUE THREAD

**Issue by mkalcok** [juju/python-libjuju#618]
Hi, would it be possible to implement `__eq__` operator for the `Unit` objects? Currently, running something like this
```python
model.units["ceph-osd/0"] == model.units["ceph-osd/0"]
```
returns `False`.

**Comment by cderici** [comment]
It would be useful, but it's actually non-trivial to implement such an equality that will work well for everyone, because depending on when it is called, the `delta`s on the units will differ, which makes the units effectively different from one another from libjuju's perspective.

The most we can provide is something that takes as input some `Unit` fields the user wants to compare and check equality based on that, but it would be no different than what you must be currently doing to check if the units are equal in the first place.

For these reasons we'll close this. It's a good suggestion, thank you for that, but unfortunately not very feasible at the moment.

**Comment by mkalcok** [comment]
I see, fair point. Thanks for the clarification.

**Comment by auria** [comment]
hi @cderici , could you give an example of `delta`s on the units and how they would differ?

From an operational perspective, a Juju model will have applications (and units, and relations). Each unit name (or unit as a whole) is unique. If a unit is removed and a new one created, we will move from having e.g. `ceph-osd/0`, `ceph-osd/1`, `ceph-osd/2` to `ceph-osd/0`, `ceph-osd/2`, `ceph-osd/3`.

In order to have 2 different `ceph-osd/1` units, we would need to list units from different models or units created after destroying the environment and creating a new model with the same model name.

In operations, I think it is understood that units may change their state (`active/error/blocked`, or new configs). I think the common expectation is to verify the unit (same unit name, same model and controller, same machine where it runs, same charm used, and maybe same revision of the charm, etc) are "the same"... I think it is well know in the Juju world that units states (in terms of configs applied) may be different if you run "juju status" today (or "juju config") and next week, but that may not be the point of the comparison.

The approach mentioned to compare a subset of attributes (name, model, controller, machine, anything else?) will probably cover the needs of someone that runs that check manually nowadays.

The `__eq__` from an application could demand more checks (like the "juju config" is equal), but that should probably not be needed at the unit level?

I'm very interested in hearing the thoughts of the Juju team on this. Thank you.

**Comment by SimonRichardson** [comment]
The API interaction between pylibjuju and juju is more akin to an internal caching layer. The way pylibjuju works is that when you connect to juju, it makes an API request to the AllWatcher API. The AllWatcher API returns a stream of deltas (see the [translation phase](https://github.com/juju/juju/blob/develop/apiserver/watcher.go#L402-L428)) of a given entity (model, application, machine, units etc). These deltas are streamed via a websocket to any pylibjuju open connection. These deltas not only contain specific fixed things like model-uuid, name, application but also things likely to change in day to day running operations; ports, workload status, agent status in case of a unit. This is for all entities.

So how does a unit __eq__ not __eq__ another unit, it depends on what your equality means? Does it mean that the name is the same or does it mean that the status of a unit is the same?

As the delta is change behind the scenes, even doing `model.units['app/0'] == model.units['app/0']` might not be correct.

> The approach mentioned to compare a subset of attributes (name, model, controller, machine, anything else?) will probably cover the needs of someone that runs that check manually nowadays.

What works for you might not work for others, as a library, I don't think this is the right approach. What we should do is offer a method for equality that takes a predicate that allows you as a library user to state what you want.

```
unit0.equals(unit1, lambda a, b: a.name == b.name)
```

To summarise we are of the shortcomings of this library and we are having a lot of discussions on how to improve or re-write it to better serve the users.

**Comment by auria** [comment]
Thanks for the detailed explanation.

As an operator, I want to know whether 2 Unit objects reference the same Juju unit. I understand that, operationally, states will change, but objects will remain running (and managing) the same charm logic, on the same Juju machine.

When automating operations, we first identify units and we then check their states. This is why `__eq__` means "same object" but not "same state" (we run other checks for this - e.g. if a unit is not healthy [error/blocked/waiting], or if a Juju action returns an unhealthy state [e.g. a Ceph cluster in WARN state], we abort further executions).
